# Meta

## Set up environment

In [2]:
!pip install datasets

INFO: pip is looking at multiple versions of multiprocess to determine which version is compatible with other requirements. This could take a while.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 471.6/471.6 kB 26.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 11.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 12.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 16.6 MB/s eta 0:00:00


In [1]:
import torch as t
import torch.nn as nn

### Github Sync

In [5]:
# Github repo
# !git clone https://github.com/hezhonghao/InstructionFollowing.git
# !git clone <Personal token>




Cloning into 'InstructionFollowing'...
remote: Enumerating objects: 6, done.
remote: Counting objects: 100% (6/6), done.
remote: Compressing objects: 100% (5/5), done.
remote: Total 6 (delta 0), reused 0 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (6/6), 4.88 KiB | 4.88 MiB/s, done.


In [6]:
%cd /content/InstructionFollowing

/content/InstructionFollowing


In [7]:
!pwd

/content/InstructionFollowing


In [ ]:
!git rm --cached InstructionFollowing


rm 'InstructionFollowing'


In [ ]:
# Forget stored credentials
!git config --system --unset credential.helper
!git config --global --unset credential.helper
!git config --local --unset credential.helper

In [ ]:
!git config user.name
!git config user.email

hezhonghao
thisisalexhe@gmail.com


In [ ]:
# Reset cached credentials (if any)
!git credential-cache exit

!git config --global user.email "thisisalexhe@gmail.com"
!git config --global user.name "hezhonghao"


# Use the token in place of your password when pushing
# !git remote set-url origin <Github personal access token>@github.com/hezhonghao/InstructionFollowing.git


In [ ]:
!git push origin main

remote: Permission to hezhonghao/InstructionFollowing.git denied to hezhonghao.
fatal: unable to access 'https://github.com/hezhonghao/InstructionFollowing.git/': The requested URL returned error: 403


In [ ]:
# Update remote URL to include your username
!git remote set-url origin https://hezhonghao@github.com/hezhonghao/InstructionFollowing.git

In [ ]:
# Github sync
!git add .
!git commit -m "Q&A baseline eval w/ squad"
!git push origin main

hint: You've added another git repository inside your current repository.
hint: Clones of the outer repository will not contain the contents of
hint: the embedded repository and will not know how to obtain it.
hint: If you meant to add a submodule, use:
hint: 
hint: 	git submodule add <url> InstructionFollowing
hint: 
hint: If you added this path by mistake, you can remove it from the
hint: index with:
hint: 
hint: 	git rm --cached InstructionFollowing
hint: 
hint: See "git help submodule" for more information.
On branch main
Your branch is ahead of 'origin/main' by 2 commits.
  (use "git push" to publish your local commits)

nothing to commit, working tree clean
fatal: could not read Password for 'https://hezhonghao@github.com': No such device or address


In [ ]:
import os
os.chdir('/content/InstructionFollowing')

In [ ]:
# Python can search Modules if you add your repo in this path (and modules in this repo)
import sys
sys.path.append('/content/InstructionFollowing')  # Adjust this path if necessary


# The Models

NEP I am actually not too sure how did they decide what arguments when you instantiate this function and what arguments when you call this instance.

In [2]:
from transformers import pipeline, set_seed
generator = pipeline('text-generation', model='gpt2')
set_seed(42)
generator("Hello, I'm a Donal Ttrump,", max_length=100, num_return_sequences=5)


/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(
Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': 'Hello, I\'m a Donal Ttrump, but what I do in that role is different—I\'ve been here for nearly 12 years, and I\'ve heard what people were saying: \'Well, your whole thing is so stupid, this isn\'t your game. The first thing to do is show people that you can still be good, and people will say, \'What?\' It\'s the hard part again."\n\nThe first thing to do is draw a big line.\n\n"'},
 {'generated_text': "Hello, I'm a Donal Ttrump, this is my first public and I've always wanted to get into wrestling. I've never been to a show before, my friends over at WNYC, I've never seen it.\n\nI'll tell you this is the first I've watched wrestling under Jim Ross. This is my first PWG show, and, by the way, not my first.\n\nI've seen wrestling now and I was probably the first one to know"},
 {'generated_text': "Hello, I'm a Donal Ttrump, and I'll do anything in my power to help you out!\n\nHARRY NOMINEE-WAN: Welcome to my show. Thanks for joining. I hope I'm not the only one who kn

In [ ]:
generator("Hello I am Elon babe,", max_length=100, num_return_sequences=5)

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': "Hello I am Elon babe, I am also the first to admit that I am absolutely sick of reading all this stuff about sex and being a sexual asshole. It's simply not right to be a sexual asshole and yet I can get a hold of it. Just kidding. I know exactly what you mean. And fuck all you say to me because I have my own opinions on it. No need to get into specifics because I promise you we will keep discussing it in the next couple of days when I"},
 {'generated_text': 'Hello I am Elon babe, and I\'m talking about, what\'s your nickname for this weekend: J-chan? I\'m kind of obsessed about this chick like everyone loves girls from a younger generation. I\'m also interested to hear how you feel about men and the world before they get a taste of their true selves.\n\n"Oh my gosh what a great job I have now that my ex-boyfriend has got to make his own way into porn. Well then that\'s where'},
 {'generated_text': "Hello I am Elon babe, you're making a difference by standing up fo

In [15]:
generator("When was the Empire State Building completed?", max_length=100, num_return_sequences=5)

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': 'When was the Empire State Building completed? What about the Capitol Building and the World Trade Center?\n\nHere in San Diego, they have the answer.\n\nIn 2005, San Diego City Council passed the San Diego Tax Relief Act (STRA) to help the people of San Diego be able to benefit from new tax breaks. This legislation included a provision which would have allowed more wealthy San Diego taxpayers to deduct their state and local taxes paid as taxes, when they received federal funds in exchange for'},
 {'generated_text': 'When was the Empire State Building completed?\n\nI\'m glad, as the Empire State Building was not completed until 1978, we had no real reason to go to the next level. It\'s a long time before we finally got to an area of the structure—what can the next one be up to?\n\nWhy are you calling it a "new era of luxury?"\n\nWell, it\'s true that the Empire State Building is not finished. It does not have an original plan of'},
 {'generated_text': 'When was the 

How to use this model to get the features of a given text in PyTorch:

In [ ]:
from transformers import GPT2Tokenizer, GPT2Model
tokenizer = GPT2Tokenizer.from_pretrained('gpt2')
model = GPT2Model.from_pretrained('gpt2')  # define the pre-trained model used in this pipeline
text = "Replace me by any text you'd like."
encoded_input = tokenizer(text, return_tensors='pt')
output = model(**encoded_input)
output

BaseModelOutputWithPastAndCrossAttentions(last_hidden_state=tensor([[[ 0.1629, -0.2166, -0.1410,  ..., -0.2619, -0.0819,  0.0092],
         [ 0.4628,  0.0248, -0.0785,  ..., -0.0859,  0.5122, -0.3939],
         [-0.0644,  0.1551, -0.6306,  ...,  0.2488,  0.3691,  0.0833],
         ...,
         [-0.5591, -0.4490, -1.4540,  ...,  0.1650, -0.1302, -0.3740],
         [ 0.1400, -0.3875, -0.7916,  ..., -0.1780,  0.1824,  0.2185],
         [ 0.1721, -0.2420, -0.1124,  ..., -0.1068,  0.1205, -0.3213]]],
       grad_fn=<ViewBackward0>), past_key_values=((tensor([[[[-1.0719,  2.4170,  0.9660,  ..., -0.4787, -0.3316,  1.7925],
          [-2.2897,  2.5424,  0.8317,  ..., -0.5299, -2.4828,  1.3537],
          [-2.2856,  2.7125,  2.4725,  ..., -1.4911, -1.8427,  1.6493],
          ...,
          [-3.3203,  2.3325,  2.7061,  ..., -1.1569, -1.5586,  2.4076],
          [-2.9917,  2.2701,  2.1742,  ..., -0.8670, -1.6410,  1.9237],
          [-2.5066,  2.6140,  2.1347,  ..., -0.0627, -2.0542,  1.6568]],

This model has bias (training data not released)

In [ ]:
generator("The White man worked as a", max_length=10, num_return_sequences=5)

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': 'The White man worked as a contractor for a company'},
 {'generated_text': 'The White man worked as a cook for a large'},
 {'generated_text': 'The White man worked as a cook at a farm'},
 {'generated_text': 'The White man worked as a security guard, and'},
 {'generated_text': 'The White man worked as a secretary or was a'}]

In [ ]:
generator("The Black man worked as a", max_length=10, num_return_sequences=5)

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': 'The Black man worked as a cook at the shop'},
 {'generated_text': 'The Black man worked as a garda, but'},
 {'generated_text': 'The Black man worked as a security guard for a'},
 {'generated_text': 'The Black man worked as a carpenter.\n'},
 {'generated_text': 'The Black man worked as a cop in Houston for'}]

In [ ]:
generator("The Asian man worked as a", max_length=10, num_return_sequences=5)

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': 'The Asian man worked as a janitor for the'},
 {'generated_text': 'The Asian man worked as a manager at a Chinese'},
 {'generated_text': 'The Asian man worked as a police negotiator for 11'},
 {'generated_text': 'The Asian man worked as a barber and the'},
 {'generated_text': 'The Asian man worked as a waiter at a Chinese'}]

In [ ]:
generator("The Arab man worked as a", max_length=10, num_return_sequences=5)

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': 'The Arab man worked as a waiter and the manager'},
 {'generated_text': 'The Arab man worked as a taxi driver in the'},
 {'generated_text': 'The Arab man worked as a driver there, a'},
 {'generated_text': 'The Arab man worked as a baker, and one'},
 {'generated_text': 'The Arab man worked as a guard at an outpost'}]

# Fine-tune

## Eval language modeling on Wikitext-103
Steps: <br>
1/ Get the model <br>
2/ Get a proportion of wikitext test  <br>
3/ test its perplexity

In [ ]:
# Loading model
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM
from datasets import load_dataset
import math

# Load the model and tokenizer
model_name = "gpt2"  # You can change this to any model you want to evaluate
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)
model.eval()

# Load Wikitext-103 dataset
dataset = load_dataset("wikitext", "wikitext-103-v1", split="test")


In [ ]:
# A demonstration of language modeling tasks on wikitext-103 # TODO


In [20]:

def calculate_perplexity(model, tokenizer, text):
    encodings = tokenizer(text, return_tensors="pt")
    max_length = model.config.max_position_embeddings
    stride = 512
    seq_len = encodings.input_ids.size(1)

    nlls = []
    prev_end_loc = 0
    for begin_loc in range(0, seq_len, stride):
        end_loc = min(begin_loc + max_length, seq_len)
        trg_len = end_loc - prev_end_loc  # may be different from stride on last loop
        input_ids = encodings.input_ids[:, begin_loc:end_loc]
        target_ids = input_ids.clone()
        target_ids[:, :-trg_len] = -100

        with torch.no_grad():
            outputs = model(input_ids, labels=target_ids)
            neg_log_likelihood = outputs.loss

        nlls.append(neg_log_likelihood)

        prev_end_loc = end_loc
        if end_loc == seq_len:
            break

    ppl = torch.exp(torch.stack(nlls).mean())
    return ppl.item()

# Evaluate on a subset of Wikitext-103
num_samples = 1000  # Adjust this based on your computational resources
total_ppl = 0

for i in range(num_samples):
    text = dataset[i]["text"]
    if text.strip():  # Skip empty lines
        ppl = calculate_perplexity(model, tokenizer, text)
        total_ppl += ppl
        print(f"Sample {i+1}/{num_samples}, Perplexity: {ppl:.2f}")

average_ppl = total_ppl / num_samples
print(f"\nAverage Perplexity on Wikitext-103 (subset): {average_ppl:.2f}")

Sample 2/1000, Perplexity: 889.28
Sample 4/1000, Perplexity: 56.62
Sample 5/1000, Perplexity: 46.29
Sample 7/1000, Perplexity: 1335.94
Sample 10/1000, Perplexity: 162.06
Sample 12/1000, Perplexity: 50.77
Sample 13/1000, Perplexity: 58.70
Sample 15/1000, Perplexity: 154.73
Sample 17/1000, Perplexity: 55.27
Sample 18/1000, Perplexity: 49.64
Sample 20/1000, Perplexity: 877.32
Sample 23/1000, Perplexity: 178.09
Sample 26/1000, Perplexity: 183.65
Sample 29/1000, Perplexity: 188.17
Sample 33/1000, Perplexity: 7396.65
Sample 35/1000, Perplexity: 62.99
Sample 36/1000, Perplexity: 38.21
Sample 38/1000, Perplexity: 482.81
Sample 40/1000, Perplexity: 66.80
Sample 42/1000, Perplexity: 164.84
Sample 44/1000, Perplexity: 54.68
Sample 45/1000, Perplexity: 23.54
Sample 46/1000, Perplexity: 85.30
Sample 47/1000, Perplexity: 51.44
Sample 48/1000, Perplexity: 39.46
Sample 49/1000, Perplexity: 64.79
Sample 50/1000, Perplexity: 27.91
Sample 51/1000, Perplexity: 45.51
Sample 53/1000, Perplexity: 173.64
Samp

How does a test example look like:

= Valkyria Chronicles III =

Senjō no Valkyria 3 : Unrecorded Chronicles ( Japanese : 戦場のヴァルキュリア3 , lit . Valkyria of the Battlefield 3 ) , commonly referred to as Valkyria Chronicles III outside Japan , is a tactical role @-@ playing video game developed by Sega and Media.Vision for the PlayStation Portable . Released in January 2011 in Japan , it is the third game in the Valkyria series . Employing the same fusion of tactical and real @-@ time gameplay as its predecessors , the story runs parallel to the first game and follows the " Nameless " , a penal military unit serving the nation of Gallia during the Second Europan War .

## Fine-tune on Wikitext-103

NEP - There is a problem that you are not fine-tuning the modle that you have already tested for baseline performance. Although they are actually just the same GPT-2 model. 13

In [8]:
(claude code)
import torch
from transformers import GPT2LMHeadModel, GPT2TokenizerFast, TextDataset, DataCollatorForLanguageModeling
from transformers import Trainer, TrainingArguments
from datasets import load_dataset

# Step 1: Load the model and tokenizer
model_name = "gpt2"
model = GPT2LMHeadModel.from_pretrained(model_name)
tokenizer = GPT2TokenizerFast.from_pretrained(model_name)

# Set padding token  #NEP what is this?
tokenizer.pad_token = tokenizer.eos_token
model.config.pad_token_id = model.config.eos_token_id

# Step 2: Prepare the dataset (subset)
def tokenize_function(examples):
    return tokenizer(examples["text"], truncation=True, max_length=512)

dataset = load_dataset("wikitext", "wikitext-103-v1")
train_subset = dataset["train"].select(range(10000))  # Use only first 10,000 examples
tokenized_dataset = train_subset.map(tokenize_function, batched=True, remove_columns=["text"])


# Step 3: Create data collator
data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm=False)


# Step 4: Define training arguments (modified)
training_args = TrainingArguments(
    output_dir="./gpt2-quick-finetuned-wikitext103",
    overwrite_output_dir=True,
    num_train_epochs=1,  # Keep this at 1 for quick training
    per_device_train_batch_size=8,  # Increase if you have more GPU memory
    save_steps=1000,
    save_total_limit=2,
    logging_dir="./logs",
    logging_steps=100,
    learning_rate=5e-4,  # Slightly higher learning rate
    warmup_steps=100,
    weight_decay=0.01,  # Add some regularization
    fp16=True if torch.cuda.is_available() else False,  # Use mixed precision training if your GPU supports it
)

# Step 5: Create Trainer
trainer = Trainer(
    model=model,
    args=training_args,
    data_collator=data_collator,
    train_dataset=tokenized_dataset,
)



# Step 6: Fine-tune the model
trainer.train()

# Step 7: Save the fine-tuned model
model.save_pretrained("./gpt2-finetuned-wikitext103")
tokenizer.save_pretrained("./gpt2-finetuned-wikitext103")

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


README.md:   0%|          | 0.00/10.5k [00:00<?, ?B/s]

test-00000-of-00001.parquet:   0%|          | 0.00/722k [00:00<?, ?B/s]

train-00000-of-00002.parquet:   0%|          | 0.00/156M [00:00<?, ?B/s]

train-00001-of-00002.parquet:   0%|          | 0.00/156M [00:00<?, ?B/s]

validation-00000-of-00001.parquet:   0%|          | 0.00/655k [00:00<?, ?B/s]

Generating test split:   0%|          | 0/4358 [00:00<?, ? examples/s]

Generating train split:   0%|          | 0/1801350 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/3760 [00:00<?, ? examples/s]

Map:   0%|          | 0/10000 [00:00<?, ? examples/s]

/usr/local/lib/python3.10/dist-packages/accelerate/accelerator.py:494: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  self.scaler = torch.cuda.amp.GradScaler(**kwargs)


We strongly recommend passing in an `attention_mask` since your input_ids may be padded. See https://huggingface.co/docs/transformers/troubleshooting#incorrect-output-when-padding-tokens-arent-masked.
You may ignore this warning if your `pad_token_id` (50256) is identical to the `bos_token_id` (50256), `eos_token_id` (50256), or the `sep_token_id` (None), and your input is not padded.


OutOfMemoryError: CUDA out of memory. Tried to allocate 8.86 GiB. GPU 0 has a total capacity of 39.56 GiB of which 6.29 GiB is free. Process 22909 has 33.27 GiB memory in use. Of the allocated memory 25.75 GiB is allocated by PyTorch, and 7.02 GiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)

In [10]:
# Step 8 evaluation of fine-tuned model  (claude code)
from tqdm import tqdm

def calculate_perplexity(model, tokenizer, dataset, max_length=512, batch_size=1):
    model.eval()
    total_loss = 0
    total_length = 0

    device = 'cuda' if torch.cuda.is_available() else 'cpu'
    model.to(device)

    with torch.no_grad():
        for i in tqdm(range(0, len(dataset), batch_size), desc="Calculating Perplexity"):
            batch = dataset[i:i+batch_size]

            # Check if the batch has any content
            if not batch['text']:
                continue

            encodings = tokenizer(batch["text"], truncation=True, max_length=max_length, return_tensors="pt", padding=True)
            input_ids = encodings.input_ids.to(device)

            # Check if input_ids is not empty
            if input_ids.numel() == 0:
                continue

            outputs = model(input_ids, labels=input_ids)

            total_loss += outputs.loss.item() * input_ids.size(1)
            total_length += input_ids.size(1)

            # Clear GPU cache
            if device == 'cuda':
                torch.cuda.empty_cache()

    if total_length == 0:
        raise ValueError("No valid inputs were processed. Check your dataset and tokenization.")

    avg_loss = total_loss / total_length
    perplexity = torch.exp(torch.tensor(avg_loss))
    return perplexity.item()

# Usage:
test_subset = dataset["test"].select(range(1000))  # Use only first 1,000 test examples
try:
    test_perplexity = calculate_perplexity(model, tokenizer, test_subset, batch_size=1)
    print(f"Perplexity on test subset: {test_perplexity:.2f}")
except ValueError as e:
    print(f"Error: {e}")
    print("Please check your dataset and ensure it contains valid text data.")

Calculating Perplexity: 100%|██████████| 1000/1000 [00:09<00:00, 102.43it/s]

Perplexity on test subset: 34.66


### Report
1/ Our fine-tuned model has acheived perplexity 34.66, compared published baseline of 37.5 for GPT-2 (117M) model.

2/ Base model perplexity 337.76, a lil too high. Reason unknown, yet. TODO: to investigate.

# The Data

# The Tasks

# The Evaluations

# The Interps